src/alternative-gas-token/L1GasTokenGateway.sol

```diff
@@ -0,0 +1,186 @@
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /**********
+     * Errors *
+     **********/
+
+    /// @dev Thrown when `msg.value` is not zero.
+    error ErrorNonZeroMsgValue();
+
+    /// @dev Thrown when the selector is invalid during `onDropMessage`.
+    error ErrorInvalidSelector();
+
+    /// @dev Thrown when the deposit amount is zero.
+    error ErrorDepositZeroGasToken();
+
+    /*************
+     * Constants *
+     *************/
+
+    /// @dev The address of gas token.
+    address public immutable gasToken;
+
+    /// @dev The scalar to scale the gas token decimals to 18.
+    uint256 public immutable scale;
+
+    /***************
+     * Constructor *
+     ***************/
+
+    /// @notice Constructor for `L1GasTokenGateway` implementation contract.
+    ///
+    /// @param _gasToken The address of gas token in L1.
+    /// @param _counterpart The address of `L2ETHGateway` contract in L2.
+    /// @param _router The address of `L1GatewayRouter` contract in L1.
+    /// @param _messenger The address of `L1ScrollMessenger` contract in L1.
+    constructor(
+        address _gasToken,
+        address _counterpart,
+        address _router,
+        address _messenger
+    ) ScrollGatewayBase(_counterpart, _router, _messenger) {
+        if (_gasToken == address(0) || _router == address(0)) revert ErrorZeroAddress();
+
+        _disableInitializers();
+
+        gasToken = _gasToken;
+        scale = 10**(18 - IERC20MetadataUpgradeable(_gasToken).decimals());
+    }
+
+    /// @notice Initialize the storage of L1GasTokenGateway.
+    function initialize() external initializer {
+        ScrollGatewayBase._initialize(address(0), address(0), address(0));
+    }
+
+    /*****************************
+     * Public Mutating Functions *
+     *****************************/
+
+    /// @inheritdoc IL1ETHGateway
+    function depositETH(uint256 _amount, uint256 _gasLimit) external payable override {
+        _deposit(_msgSender(), _amount, new bytes(0), _gasLimit);
+    }
+
+    /// @inheritdoc IL1ETHGateway
+    function depositETH(
+        address _to,
+        uint256 _amount,
+        uint256 _gasLimit
+    ) external payable override {
+        _deposit(_to, _amount, new bytes(0), _gasLimit);
+    }
+
+    /// @inheritdoc IL1ETHGateway
+    function depositETHAndCall(
+        address _to,
+        uint256 _amount,
+        bytes calldata _data,
+        uint256 _gasLimit
+    ) external payable override {
+        _deposit(_to, _amount, _data, _gasLimit);
+    }
+
+    /// @inheritdoc IL1ETHGateway
+    function finalizeWithdrawETH(
+        address _from,
+        address _to,
+        uint256 _amount,
+        bytes calldata _data
+    ) external payable override onlyCallByCounterpart nonReentrant {
+        if (msg.value > 0) {
+            revert ErrorNonZeroMsgValue();
+        }
+
+        uint256 downScaledAmount = _amount / scale;
```

Comment:
I'm slightly concerned about this downscaling.

What if `l1_decimals = 6`, but I withdraw `N + 1/1e6 / 2` tokens 2 times?

Then we'd have `1/1e6` less tokens on L2 compared to L1, i.e. we'd have some dust (`1/1e6` tokens) stuck in the L1 gateway forever.

Maybe not an issue, just mentioning.

Reply:
Conclusion: it is ok to left some dust tokens. Users usually won’t care about this. and the invariant that "locked on L1 >= released on L2" still holds.